subsys/net/openthread/rpc/server/ot_rpc_ctrl_server.c

```diff
@@ -544,6 +547,142 @@ static void ot_rpc_cmd_get_factory_assigned_eui64(const struct nrf_rpc_group *gr
 	nrf_rpc_cbor_rsp_no_err(group, &rsp_ctx);
 }
 
+static void ot_rpc_cmd_netdata_get(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,
+				   void *handler_data)
+{
+	struct nrf_rpc_cbor_ctx rsp_ctx;
+	otError error = OT_ERROR_NONE;
+	bool stable;
+	bool decoded_ok;
+	uint8_t length = 0;
+
+	NRF_RPC_CBOR_ALLOC(group, rsp_ctx, sizeof(error) + length + 3);
+
+	decoded_ok = zcbor_bool_decode(ctx->zs, &stable);
+
+	if (!decoded_ok) {
+		goto exit;
+	}
+
+	decoded_ok = zcbor_uint_decode(ctx->zs, &length, sizeof(length));
+	if (!decoded_ok) {
+		goto exit;
+	}
+
+	nrf_rpc_cbor_decoding_done(group, ctx);
+
+	if (!zcbor_bstr_start_encode(rsp_ctx.zs)) {
+		goto exit;
+	}
+
+	openthread_api_mutex_lock(openthread_get_default_context());
+
+	error = otNetDataGet(openthread_get_default_instance(), stable, rsp_ctx.zs[0].payload_mut,
+			     &length);
+
+	openthread_api_mutex_unlock(openthread_get_default_context());
+
+	if (!zcbor_bstr_end_encode(rsp_ctx.zs, NULL)) {
+		goto exit;
+	}
+
+exit:
+	if (!decoded_ok) {
+		nrf_rpc_cbor_decoding_done(group, ctx);
+	}
+
+	zcbor_uint_encode(rsp_ctx.zs, &error, sizeof(error));
+	nrf_rpc_cbor_rsp_no_err(group, &rsp_ctx);
+}
+
+static void ot_rpc_cmd_get_next_service(const struct nrf_rpc_group *group,
+					struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	struct nrf_rpc_cbor_ctx rsp_ctx;
+	otError error = OT_ERROR_NONE;
+	otNetworkDataIterator iterator;
+	bool decoded_ok = false;
+	otServiceConfig service_config;
+
+	NRF_RPC_CBOR_ALLOC(group, rsp_ctx, sizeof(error) + sizeof(service_config) + 7);
+
+	decoded_ok = zcbor_uint_decode(ctx->zs, &iterator, sizeof(iterator));
+
+	if (!decoded_ok) {
+		error = OT_ERROR_INVALID_ARGS;
+		goto exit;
+	}
+
+	nrf_rpc_cbor_decoding_done(group, ctx);
+
+	openthread_api_mutex_lock(openthread_get_default_context());
+	error = otNetDataGetNextService(openthread_get_default_instance(), &iterator,
+					&service_config);
+	openthread_api_mutex_unlock(openthread_get_default_context());
+
+	if (!zcbor_uint_encode(rsp_ctx.zs, &iterator, sizeof(iterator))) {
+		error = OT_ERROR_FAILED;
+		goto exit;
+	}
+
+	if (error == OT_ERROR_NONE) {
+		ot_rpc_encode_service_config(&rsp_ctx, &service_config);
+	} else {
+		ot_rpc_encode_service_config(&rsp_ctx, NULL);
+	}
+
+exit:
+	if (!decoded_ok) {
+		nrf_rpc_cbor_decoding_done(group, ctx);
+	}
+	zcbor_uint_encode(rsp_ctx.zs, &error, sizeof(error));
```

Comment:
If we always encode the error, wouldn't it make sense to put in at the beginning so that the decoder can use it to know whether to go on decoding the service config?